demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,441 @@
+    if( ( status == false ) && ( pTopicName[ 0 ] != '$' ) )
+    {
+        status = topicFilterMatch( pTopicName, topicNameLength, pTopicFilter, topicFilterLength );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/* Design -
+ *   * Common handler can contain logic of processing acks and retries.
+ *   * Only forward PUBLISH message to message dispatcher/handler as individual callbacks
+ *     only want to process incoming PUBLISH message.
+ */
+void SubscriptionManager_DispatchHandler( MQTTContext_t * pContext,
+                                          MQTTPublishInfo_t * pPublishInfo )
+{
+    assert( pPublishInfo != NULL );
+    assert( pContext != NULL );
+
+    size_t listIndex = 0u;
+
+    /* Iterate through record list to find matching topics, and invoke their callbacks. */
+    for( listIndex = 0; listIndex < MAX_SUBSCRIPTION_CALLBACK_RECORDS; listIndex++ )
+    {
+        if( ( callbackRecordList[ listIndex ].pTopicFilter != NULL ) &&
+            ( matchTopic( pPublishInfo->pTopicName,
+                          pPublishInfo->topicNameLength,
+                          callbackRecordList[ listIndex ].pTopicFilter,
+                          callbackRecordList[ listIndex ].topicFilterLength ) == true ) )
+        {
+            LogInfo( ( "Invoking subscription callback of matching topic filter: "
+                       "TopicFilter=%.*s, TopicName=%.*s",
+                       callbackRecordList[ listIndex ].topicFilterLength,
+                       callbackRecordList[ listIndex ].pTopicFilter,
+                       pPublishInfo->topicNameLength,
+                       pPublishInfo->pTopicName ) );
+
+            /* Invoke the callback associated with the record as the topics match. */
+            callbackRecordList[ listIndex ].callback( pContext, pPublishInfo );
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+bool SubscriptionManager_RegisterCallback( const char * pTopicFilter,
+                                           uint16_t topicFilterLength,
+                                           SubscriptionManager_Callback_t callback )
+{
+    assert( pTopicFilter != NULL );
+    assert( topicFilterLength != 0 );
+
+    bool recordAdded = false;
+    size_t availableIndex = 0u;
+    bool recordExists = false;
+
+    /* Search for an available spot in the list to store the record */
+    while( ( availableIndex < MAX_SUBSCRIPTION_CALLBACK_RECORDS ) &&
+           ( callbackRecordList[ availableIndex ].pTopicFilter != NULL ) )
+    {
+        /* Check if a record for the topic filter already exists in the registry. */
+        if( ( callbackRecordList[ availableIndex ].topicFilterLength == topicFilterLength ) &&
+            ( strncmp( pTopicFilter, callbackRecordList[ availableIndex ].pTopicFilter, topicFilterLength )
+              == 0 ) )
+        {
+            recordExists = true;
```

Comment:
I think we mentioned this in the other PR, but we ought to document that this doesn't allow multiple callbacks for the same topic filter. I think it's plausible that more than one application wants to receive messages from the same topic filter, in which case we should point to the code that needs to be changed in order to allow it.

Reply:
Added a `@note` in the API documentation of the function in the header file

Reply:
I don't think that adding documentation for supporting multiple callback registration per topic filter is necessarily required. (The source code is available for viewing and tweaking). 
If we really think  that there can be a customer need for multiple callbacks per topic filter, then we should just support that in the subscription manager. From the perspective of a single threaded application, I don;t see a reason why multiple callbacks per topic filter would be needed.